_posts/2025-02-04-data-wants-to-be-free.md

```diff
@@ -0,0 +1,306 @@
+00000018: <span class="a-length">08</span> <span class="a-data">00 00 00 00 00 00 00</span>  <span class="a-length">.</span><span class="a-data">.......</span>  <span class="a-length">Length of value</span>
+00000020: <span class="a-data">01</span> 00 00 00 03 66 6f 6f  <span class="a-data">.</span>....foo  <span class="a-data">Data</span>
+00000028: 00 00 00 08 00 00 00 00  ........
+00000030: 00 00 00 40 00 03 00 00  ...@....
+00000038: 00 08 00 00 00 00 00 00  ........
+00000040: 00 02 00 00 00 0f 61 20  ......a
+00000048: 6c 6f 6e 67 65 72 20 73  longer s
+00000050: 74 72 69 6e 67 00 00 00  tring...
+00000058: 08 00 00 00 00 00 00 00  ........
+00000060: 80 00 03 00 00 00 08 00  ........
+00000068: 00 00 00 00 00 00 03 00  ........
+00000070: 00 00 12 79 65 74 20 61  ...yet a
+00000078: 6e 6f 74 68 65 72 20 73  nother s
+00000080: 74 72 69 6e 67 00 00 00  tring...
+00000088: 08 00 00 00 00 00 00 00  ........
+00000090: 0a ff ff                 ...</code></pre></div></div>
+
+Honestly, PostgreSQL’s binary format is quite understandable, and pretty compact at first glance. But a closer look isn’t so favorable. **PostgreSQL has overheads proportional to the number of rows and columns**:
+
+* Every row has a 2 byte prefix for the number of values in the row. *But the data is tabular—we already know this info, and it doesn’t change\!*
+* Every value of every row has a 4 byte prefix for the length of the following data, or \-1 if the value is NULL. *But we know the data types, and those don’t change—most values have a fixed, known length\!*
+* All values are big-endian. *But most of our devices are little-endian, so the data has to be converted.*
+
+Sure, we need to store if a value is NULL or not, but 4 bytes is a *bit* much for a boolean. String data and other non-fixed-length types need per-value lengths, but PostgreSQL adds the length for *every* type of value. And converting big-endian to little-endian is pretty trivial…but it’s still work that stands in between you and your data. To PostgreSQL’s credit, its format is at least cheap and easy to parse—[other formats](https://protobuf.dev/programming-guides/encoding/) get fancy with tricks like “varint” encoding which are quite expensive.
+
+For example, a single column of int32 values would have 4 bytes of data and 6 bytes of overhead per row—**60% is “wasted\!”**[^1] The ratio gets a little better with more columns (but not with more rows); in the limit we approach “only” 50% overhead.
+
+How does Arrow compare? We can use [ADBC](https://arrow.apache.org/adbc/current/driver/postgresql.html) to pull the PostgreSQL table into an Arrow table, then annotate it like before:
+
+```console
+>>> import adbc_driver_postgresql.dbapi
+>>> import pyarrow.feather
+>>> conn = adbc_driver_postgresql.dbapi.connect("...")
+>>> cur = conn.cursor()
+>>> cur.execute("SELECT * FROM demo")
+>>> data = cur.fetchallarrow()
+>>> pyarrow.feather.write_feather(data, "demo.arrow")
```

Comment:
Is it maybe better if we use the non-Feather-branded APIs here? Did you pick this because the Feather API is more ergonomic than the IPC File API? I know we have open issues and mailing list posts about doing something about the Feather name. If you keep the Feather APIs, maybe add a note about what Feather is.

Reply:
I agree we should avoid the feather terminology

Reply:
It's much easier than remembering exactly which class. The feather/csv/parquet APIs are all (almost) consistent then IPC it's completely different. Why make the right thing the hardest API?

Reply:
(Also I don't get a deprecation notice, so as far as I'm concerned the API is still valid?)

Reply:
Totally valid, yes. My point was only about having not introduced the word "feather" before using it here. Ideally we'd renamed the API at this point.

I'm fine with an aside or comment about it but if you feel fine with it as is, go with it.

Reply:
I can change the API or add a note but if we feel the current code is wrong I think upstream should be fixed too

Reply:
We have a discussion thread for "Feather": https://lists.apache.org/thread/d2l7179yd41gfdhktrl1h06gh31l8l4l

It seems that we want to deprecate "Feather" but the discussion wasn't completed...

How about avoiding using Feather API here and adding consistent IPC API to PyArrow? (We may need to use inconsistent IPC API here for now but we may be able to update it later.)

Reply:
https://github.com/apache/arrow/issues/38515 already existed so made a note there.